src/core/__tests__/ReactErrorBoundaries-test.js

```diff
@@ -236,4 +236,353 @@ describe('ReactErrorBoundaries', function() {
+      render() {
+        return <BrokenUnmount />;
+      }
+    }
+
+    class BrokenUnmount extends React.Component {
+      render() {
+        return <div>text</div>;
+      }
+      componentWillUnmount() {
+        throw new Error('Always broken.');
+      }
+    }
+
+    var container = document.createElement('div');
+    ReactDOM.render(<ErrorBoundary angry={false} />, container);
+    ReactDOM.render(<ErrorBoundary angry={true} />, container);
+    expect(container.textContent).toBe('Error: I am now a sad component :(');
+    ReactDOM.unmountComponentAtNode(container);
+  });
+
+  it('doesn\'t get into inconsistent state during removals', function() {
+    class ErrorBoundary extends React.Component {
+      constructor(props) {
+        super(props);
+        this.state = {errorMessage: null};
+      }
+      render() {
+        if (this.state.errorMessage != null) {
+          return <div>Error: I am now a sad component :(</div>;
+        }
+
+        return (
+          <div>{this.props.children}</div>
+        );
+      }
+      unstable_handleError(e) {
+        this.setState({errorMessage: e.message});
+      }
+    }
+
+    class InnocentComponent extends React.Component {
+      render() {
+        return <div>text</div>;
+      }
+    }
+
+    class GuiltyComponent extends React.Component {
+      render() {
+        return <div>text</div>;
+      }
+      componentWillUnmount() {
+        throw new Error('Always broken.');
+      }
+    }
+
+    var container = document.createElement('div');
+    ReactDOM.render(<ErrorBoundary><InnocentComponent /><GuiltyComponent /><InnocentComponent /></ErrorBoundary>, container);
+    ReactDOM.render(<ErrorBoundary />, container);
+    expect(container.textContent).toBe('Error: I am now a sad component :(');
+    ReactDOM.unmountComponentAtNode(container);
+  });
+
+  it('doesn\'t get into inconsistent state during additions', function() {
+    class ErrorBoundary extends React.Component {
+      constructor(props) {
+        super(props);
+        this.state = {errorMessage: null};
+      }
+      render() {
+        if (this.state.errorMessage != null) {
+          return <div>Error: I am now a sad component :(</div>;
+        }
+
+        return (
+          <div>{this.props.children}</div>
+        );
+      }
+      unstable_handleError(e) {
+        this.setState({errorMessage: e.message});
+      }
+    }
+
+    class InnocentComponent extends React.Component {
+      render() {
+        return <div>text</div>;
+      }
+    }
+
+    class GuiltyComponent extends React.Component {
+      render() {
+        throw new Error('Always broken.');
+      }
+    }
+
+    var container = document.createElement('div');
+    ReactDOM.render(<ErrorBoundary />, container);
+    ReactDOM.render(<ErrorBoundary><InnocentComponent /><GuiltyComponent /><InnocentComponent /></ErrorBoundary>, container);
+    expect(container.textContent).toBe('Error: I am now a sad component :(');
```

Comment:
Could we test that refs/componentDidMount are not called on any of the mounting children but that they are called for the displayed error message?